src/Uno.UI/UI/Xaml/UIElement.RoutedEvents.cs

```diff
@@ -763,23 +771,57 @@ internal bool RaiseEvent(RoutedEvent routedEvent, RoutedEventArgs args, Bubbling
 			return RaiseOnParent(routedEvent, args, parent, ctx);
 		}
 
-		private static void TrackKeyState(RoutedEvent routedEvent, RoutedEventArgs args)
+		/// <summary>
+		/// Raise a tunneling routed event starting from the root and tunneling down to this element.
+		/// </summary>
+		internal void RaiseTunnelingEvent(RoutedEvent routedEvent, RoutedEventArgs args)
 		{
-			if (args is KeyRoutedEventArgs keyArgs)
+			if (!routedEvent.Flag.IsTunnelingEvent())
+			{
+				throw new InvalidOperationException($"Event {routedEvent.Name} is not marked as a tunneling event.");
+			}
+
+			// TODO: This is just temporary workaround before proper
+			// keyboard event infrastructure is implemented everywhere
+			// (issue #6074)
+			// The key states will be tracked again in an accompanying bubbling event (e.g. KeyDown for PreviewKeyDown),
+			// but this is fine, since it's idempotent.
+			if (routedEvent.IsKeyEvent)
 			{
-				if (routedEvent == KeyDownEvent)
+				TrackKeyState(routedEvent, args);
+			}
+
+			// On WinUI, if one of the event handlers reparents this element, the tunneling still goes through the
+			// original path.
+			foreach (var p in this.GetAllParents().Reverse())
```

Comment:
Is there a cross-platform way to get the number of parents upfront? The intention is to `stackalloc`, but only `CROSSPLATFORM` has `Depth` right now.

Reply:
Actually, a `stackalloc` won't work anyway, because these aren't primitive types. We can think about it in a separate PR if it pops up while profiling.